v2/tools/mangle-test-json/main.go

```diff
@@ -0,0 +1,251 @@
+	for _, pkg := range packages {
+		tests := byPackage[pkg]
+		if len(tests) == 1 {
+			// package-only
+			continue
+		}
+
+		totalRuntime := time.Duration(0)
+		for _, t := range tests[1:] {
+			totalRuntime += t.RunTime
+		}
+
+		overallOutcome := actionSymbol(tests[0])
+		fmt.Printf("* %s `%s` (runtime %s)\n", overallOutcome, pkg, totalRuntime)
+	}
+
+	fmt.Println()
+}
+
+var maxOutputLines = 300
+
+func printDetails(packages []string, byPackage map[string][]TestRun) {
+	fmt.Printf("## Failed Test Details\n\n")
+
+	for _, pkg := range packages {
+		tests := byPackage[pkg]
+		// check package-level indicator, which will be first ("" test name):
+		if tests[0].Action != "fail" {
+			continue // no failed tests, skip
+		}
+
+		// package name as header
+		fmt.Printf("### `%s`\n\n", pkg)
+
+		// Output info on stderr
+		fmt.Fprintf(os.Stderr, "Package failed: %s\n", pkg)
+
+		for _, test := range tests[1:] {
+			// only printing failed tests
+			if test.Action == "fail" {
+				fmt.Printf("#### `%s`\n", test.Test)
+				fmt.Printf("Failed in %s\n:", test.RunTime)
+
+				trimmedOutput, output := escapeOutput(test.Output)
+				summary := "Test output"
+				if trimmedOutput {
+					summary += fmt.Sprintf(" (trimmed to last %d lines) — full details available in log", maxOutputLines)
+				}
+
+				fmt.Printf("<details><summary>%s</summary><pre>%s</pre></details>\n\n", summary, output)
+
+				// Output info on stderr, so that test failure isn’t silent on console
+				// when running `task ci`, and that full logs are available if they get trimmed
+				fmt.Fprintf(os.Stderr, "- Test failed: %s\n", test.Test)
+				fmt.Fprintln(os.Stderr, "=== TEST OUTPUT ===")
+				for _, outputLine := range test.Output {
+					fmt.Fprint(os.Stderr, outputLine) // note that line already has newline attached
+				}
+				fmt.Fprintln(os.Stderr, "=== END TEST OUTPUT ===")
+			}
+		}
+
+		fmt.Println()
+	}
+}
+
+func escapeOutput(outputs []string) (bool, string) {
+	trimmed := false
+	if len(outputs) > maxOutputLines {
+		outputs = outputs[len(outputs)-maxOutputLines:]
+		trimmed = true
+	}
+
+	result := strings.Builder{}
+	for _, output := range outputs {
+		s := output
+		s = strings.ReplaceAll(s, "&", "&amp;")
+		s = strings.ReplaceAll(s, "<", "&lt;")
+		s = strings.ReplaceAll(s, "\n", "<br>")
+		result.WriteString(s)
+	}
+
+	return trimmed, result.String()
+}
+
+func printSlowTests(byPackage map[string][]TestRun) {
+	fmt.Printf("## Longest-running tests\n\n")
+
+	allTests := []TestRun{}
+	for _, v := range byPackage {
+		for _, t := range v[1:] { // skip "" package test
+			allTests = append(allTests, t)
+		}
+	}
+
+	sort.Slice(allTests, func(i, j int) bool {
+		return allTests[i].RunTime > allTests[j].RunTime
+	})
+
+	fmt.Println("| Package | Name | Time |")
+	fmt.Println("|---------|------|-----:|")
+	for i := 0; i < min(10, len(allTests)); i += 1 {
+		test := allTests[i]
+		fmt.Printf("| `%s` | `%s` | %s |\n", test.Package, test.Test, test.RunTime)
```

Comment:
Can the test run time be better formatted to avoid overprecision?

e.g.
* 518.900052ms -> 518ms
* 1.310857311s -> 1.3ss
* 8.589789566s -> 8.5s

Reply:
Rounded all times to `ms`. Doing better requires a bunch of custom code and I don't want this “script” to become too advanced 😁

Reply:
Implemented better rounding; if over a minute we round to seconds, else milliseconds.